consensus/src/liveness/leader_reputation.rs

```diff
@@ -171,4 +176,38 @@ impl ProposerElection for LeaderReputation {
             .unwrap_err();
         self.proposers[chosen_index]
     }
+
+    fn is_valid_proposal(&self, block: &Block) -> bool {
+        block.author().map_or(false, |author| {
+            let valid = self.is_valid_proposer(author, block.round());
+            let mut already_proposed = self.already_proposed.lock().unwrap();
+            if !valid {
+                return false;
+            }
+            // detect if the leader proposes more than once in this round
+            match block.round().cmp(&already_proposed.0) {
+                Ordering::Greater => {
+                    already_proposed.0 = block.round();
+                    already_proposed.1 = HashSet::new();
+                    already_proposed.1.insert(author);
+                    true
+                }
+                Ordering::Equal => {
+                    if already_proposed.1.contains(&author) {
+                        error!(
+                            SecurityEvent::InvalidConsensusProposal,
+                            "Multiple proposals from {} for round {}",
+                            author,
+                            block.round()
+                        );
+                        false
+                    } else {
+                        already_proposed.1.insert(author);
+                        true
+                    }
+                }
+                Ordering::Less => false,
+            }
+        })
+    }
```

Comment:
why would we ever want more than one round? perhaps we just keep track of the current round and reset it if we see a more recent round and ignore earlier ones?
also do we want to keep track of the block_id?

Reply:
What if a proposer proposes a different block in the same round for instance?  Given that a proposer should only send the exact sample proposal in the same round, there is no point in accepting any other proposal from the correct round proposer if you already got one from them.

Reply:
I think we said the same thing.

Reply:
I think I misunderstood your first sentence.  If you are saying "more than one proposal from a valid proposer in a round", then yes, I agree.

Reply:
My comment probably applies more towards the question about "why would sending out a good proposal twice fail? seems a little interesting and not idempotent"

Reply:
it's implementation details, but I mean we never send the proposal more than once, I can change it to track the block id as well

Reply:
> why would we ever want more than one round? perhaps we just keep track of the current round and reset it if we see a more recent round and ignore earlier ones?

@davidiw are you trolling? I thought that's exactly what I did lol

Reply:
ah I misunderstood it too 😄

Reply:
but this assumes the valid proposer for a given round won't change which is not necessarily true since it depends on committed round.

if the node is at round 10 but committed round is at 3, the leader maybe X, but later it synchronized the last committed round to 5, the leader maybe Y even it may not be able to vote for Y.